### DISCUSSION TITLE
[Feature]: Clone a page

### DISCUSSION THREAD

**Discussion by javadsamiee** [GrapesJS/grapesjs#5448] [Roadmap Suggestions]
Clone a predefined page just with page ID. Create a new method under https://grapesjs.com/docs/api/pages.html#pages for cloning a page.

`const pageId = pageManager.get('page-id'); pageManager.clone(pageId);`

**Comment by artf** [comment]
Moved to [Roadmap Suggestions](https://github.com/GrapesJS/grapesjs/discussions/categories/roadmap-suggestions)

**Comment by bezimeni96** [comment]
Huge +1 for this feature request! A native page cloning ability is definitely needed. For anyone else looking for a solution until then, here's an approach I've put together that's working well for me.

The core idea is to use the system clipboard to temporarily store the ID of the page being copied. The "paste" action then reads this ID, creates a new blank page, and finally populates it by cloning the components from the source page.

Here’s a breakdown of the implementation, which is written within a React/TypeScript context (`useEditor` is a hook to get the GrapesJS editor instance).

---

###  The Copy Action
Instead of serializing the entire page object, the copy function just grabs the ID of the selected page and writes it to the clipboard. This is lightweight and avoids managing a large data object.

``` JavaScript
const onCopy = () => {
  // Assuming 'selected' is the currently selected page object from GrapesJS PageManager
  const pageId = selected ? selected.getId() : '';
  if (pageId) {
    navigator.clipboard.writeText(pageId);
  }
};
```

---

### The Paste Action
The paste logic is a two-step process: first, create the new page "shell," and second, copy the content over.

**1. Create the New Page**
This function reads the page ID from the clipboard, finds the original page, and then creates a new, empty page with a unique ID and a descriptive name (e.g., "Copy of Home").

The `parentId` is used here for our specific use case of hierarchical pages, but you can remove it if you have a flat page structure.

``` JavaScript
const onPaste = async () => {
  const sourcePageId = await navigator.clipboard.readText();
  const pageManager = editor.Pages;
  const sourcePage = pageManager.get(sourcePageId);

  if (!sourcePage) {
    console.error(`Page not found for ID: ${sourcePageId}`);
    return;
  }

  // Generate a unique ID for the new page
  const newPageId = `page-${uuidv4()}`;

  pageManager.add({
    id: newPageId,
    name: `Copy of ${sourcePage.getName()}`,
    // Optional: for nested pages
    // parentId: selected?.getId(), 
  });

  // Now, copy the content from the source page to the new one
  copyPageContent(sourcePageId, newPageId);
};
```


**2. Copy the Page Content**
This is the most critical part. The function iterates through the components of the source page, clones them, and adds them to the new page.

**Key Steps:**

- **Select the Target Page:** You must select the newly created page (`editor.Pages.select(targetPage)`) before adding components. Otherwise, GrapesJS will add the components to the currently active page.

- **Iterate and Clone:** Loop through the direct children of the source page's main component (`<body>`).

- **Exclude Components (Optional):** In my case, I needed to prevent shared components like a global navigation bar from being duplicated. The `if (!component.is(...))` check handles this. You can adapt this logic to exclude any component type you need.

- **Add Cloned Components:** Use `editor.addComponents()` to add the cloned component to the new page. The built-in `component.clone()` method handles the deep cloning of the component and its children.

``` JavaScript
const copyPageContent = (fromPageId, toPageId) => {
  const sourcePage = editor.Pages.get(fromPageId);
  const targetPage = editor.Pages.get(toPageId);

  if (!sourcePage || !targetPage) return;

  // IMPORTANT: Select the target page before adding components to it.
  editor.Pages.select(targetPage);

  const sourceComponents = sourcePage.getMainComponent().components();

  for (const component of sourceComponents) {
    // Optional: Skip specific components that shouldn't be duplicated (e.g., global headers/footers).
    if (component.is('your-custom-component-type-to-skip')) {
        continue;
    }
    
    // Clone the component and add it to the new page's body.
    const clonedComponent = component.clone();
    editor.addComponents(clonedComponent);
  }
};
```

---

I'm open to any feedback on this implementation, especially if there are potential edge cases or performance issues I might have missed

**Comment by artf** [comment]
We don't have a direct method for page cloning but you can do this with the current API
```js
const selectedPage = editor.Pages.getSelected();
editor.Pages.add({
  name: `${selectedPage.getName()} (Copy)`,
  component: selectedPage.getMainComponent().clone(),
}, { select: true });
```
